### PR TITLE
STR-741: Add `zizmor` and harden CI jobs against malicious attackers

### PR DESCRIPTION
**Pull request by storopoli**
## Description

Hardens CI jobs against malicious attackers and adds a new CI job to check these in the future.

### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature/Enhancement (non-breaking change which adds functionality or enhances an existing one)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Refactor
- [x] New or updated tests
- [ ] Dependency Update
- [x] Security fix

## Notes to Reviewers

Here's information on a recent [attack](https://blog.yossarian.net/2024/12/06/zizmor-ultralytics-injection) on a GitHub repository via CI template injection.
The linked post recommends the use of [`zizmor`](https://github.com/woodruffw/zizmor) to check for common CI attack vectors.

This makes two fixes:

1. Don't have the credentials being persistent in GitHub Actions artifacts with `persist-credentials: true` in all `actions/checkout` steps. See https://woodruffw.github.io/zizmor/audits/#artipacked
2. Avoid template injections by setting the untrusted input value of the expression to an intermediate environment variable. See https://woodruffw.github.io/zizmor/audits/#template-injection and https://securitylab.github.com/resources/github-actions-untrusted-input/

## Checklist

- [x] I have performed a self-review of my code.
- [x] I have commented my code where necessary.
- [x] I have updated the documentation if needed.
- [x] My changes do not introduce new warnings.
- [x] I have added tests that prove my changes are effective or that my feature works.
- [x] New and existing tests pass with my changes.

## Related Issues

STR-741.

### CODE REVIEW

**Comment by AaronFeickert** [comment]
How are the results of a `zizmor` handled? Will the CI job fail? Or are the results posted somewhere for review?

**Comment by storopoli** [comment]
> How are the results of a `zizmor` handled? Will the CI job fail? Or are the results posted somewhere for review?

I added two `zizmor` CI jobs. One will run daily on `main` and report stuff in the [Security tab](https://github.com/alpenlabs/strata/security).
Another will run on every PR and fail if find a security issue

The first one is useful since `zizmor` uses a rule set and its updated.
The second one is useful for regular PRs.

**Comment by AaronFeickert** [comment]
Nice! The action linter and `zizmor` job should probably be made required checks for merging. If it gets too noisy, we can always adjust the settings.

**Comment by storopoli** [comment]
> Nice! The action linter and `zizmor` job should probably be made required checks for merging. If it gets too noisy, we can always adjust the settings.

Yes! But I can only add after the PR has landed and has run successfully at least once in `main`

File: .github/workflows/security.yml.disabled
Comment (AaronFeickert) [inline .github/workflows/security.yml.disabled:27]: ?

File: .github/workflows/zizmor.yml
Comment (AaronFeickert) [inline .github/workflows/zizmor.yml:21]: Change to run against `.` for consistency with the automated version, perhaps.

**Review by AaronFeickert — CHANGES_REQUESTED** [review]

File: .github/workflows/dependencies.yml
Comment (AaronFeickert) [inline .github/workflows/dependencies.yml:1]: Has this been tested manually?

File: .github/workflows/security.yml.disabled
Comment (storopoli) [reply .github/workflows/security.yml.disabled:27]: Oh god autoformat on trailing whitespace....

File: .github/workflows/zizmor.yml
Comment (storopoli) [reply .github/workflows/zizmor.yml:21]: The automated runs on `.`? Ok will do. change

File: .github/workflows/dependencies.yml
Comment (storopoli) [reply .github/workflows/dependencies.yml:1]: No, but this just adds shellcheck lints.

File: .github/workflows/zizmor.yml
Comment (storopoli) [reply .github/workflows/zizmor.yml]: Done in 554a49d3130493f4040f92be39f1d8f19aed0833

File: .github/workflows/dependencies.yml
Comment (AaronFeickert) [reply .github/workflows/dependencies.yml:1]: Right, but it should definitely be tested manually after merge to ensure it works. Otherwise we need to wait until it fires and remember to check its logs manually.

Comment (storopoli) [reply .github/workflows/dependencies.yml:1]: Yes definitely it has a `workflow_dispatch:` that I can run manually.

**Review by AaronFeickert — APPROVED** [review]
LGTM, provided any automated jobs are manually tested after merge if feasible.

**Review by delbonis — APPROVED** [review]

**Comment by AaronFeickert** [comment]
The dependency update job [does not seem to work](https://github.com/alpenlabs/strata/actions/runs/12357753728/job/34486745701).

**Pull request merged**
